### ISSUE TITLE
sigsegv using mysql export

### ISSUE THREAD

**Issue by SBFspot** [SBFspot/SBFspot#84]
Today i tried to export my SMA Converter data with SBFspot into my mysql / mariadb database. Running it after midnight i got a sigsegv exeption. To my surprise bevore midnight i had no error. Maybe this should be checked, too.

But after fidling around some time with gdb i think the following line of code was causing the trouble:

for (last_rec = (sizeof(inverters[inv]->dayData)/sizeof(DayData))-1; (inverters[inv]->dayData[last_rec].watt <= 0) && (last_rec >= 0); last_rec--);

last_rec can go below zero in "inverters[inv]->dayData[last_rec].watt" when checking with "(last_rec >= 0)"

If my rusty debugging skills are right here i recommend to fix this by using "(last_rec > 0)".

As i have seen this should be checked in db_SQLite_Export.cpp, too.

#### Attachments
[db_MySQL_Export.cpp](https://www.codeplex.com/Download/AttachmentDownload.ashx?ProjectName=SBFspot&WorkItemId=79&FileAttachmentId=1357060)

#### This work item was migrated from CodePlex
CodePlex work item ID: '79'
Assigned to: 'SBF'
Vote count: '1'

**Comment by SBFspot** [comment]
[SBF@15/12/2014]
Not so sure about your explanation...
dayData is a zero indexed array[288]. With your proposal you'll never address the first element. Even more, since last_rec is unsigned int, it will never be lower than 0
``` C
unsigned int first_rec, last_rec;
for (first_rec = 0; (dayData[first_rec].watt <= 0) && (first_rec < sizeof(dayData)/sizeof(DayData)); first_rec++);
for (last_rec = (sizeof(dayData)/sizeof(DayData))-1; (dayData[last_rec].watt <= 0) && (last_rec >= 0); last_rec--);
```

**Comment by SBFspot** [comment]
[janus4@26/12/2014]
Hi,
Problem is i still there and i get SIGSEGV after calling SBFspot at night (Now it is 00:47 at 27.12.2014).

So my fix is not realy a fix but the SIGSEV is still there.

I call SBFspot using:
bash-4.3# ./bin/Release_MySQL/SBFspot -cfg/opt/bin/SBFspot.cfg -v5 -finq

.. deletet the output of -v5 ..

======
ExportDayDataToCSV()
Segmentation fault (core dumped)

Calling it without -v5:

bash-4.3# ./bin/Release_MySQL/SBFspot -cfg/opt/bin/SBFspot.cfg -finq
SBFspot V3.0.3
Yet another tool to read power production of SMA solar inverters
(c) 2012-2014, SBF (https://sbfspot.codeplex.com)
Compiled for Linux (LE) 64 bit

Commandline Args: -cfg/opt/bin/SBFspot.cfg -finq
Segmentation fault (core dumped)

I will try to compile SBFspot with debug information again and show you the result of the debugger in the next days. Maybe the problem is not really a problem as i asume it will only arise at night after midnight when no data is collectet. But it is a SIGSEGV.... :-)

The SIGSEGV will arise when the array is used for checking in the for loop and when last_rec gets an interger underrun ( last_rec = 0 - 1 ) Thats when the SIGSEGV arises for accessing the array with a far to high last_rec variable.

Greetings

Mario

**Comment by SBFspot** [comment]
[SBF@15/12/2014]
Not a bug - see my comment


** Closed by SBF 12/15/2014 12:39PM

**Comment by SBFspot** [comment]
[janus4@26/12/2014]
Forgot to mention: I Used the last SBFspot Sourcecode:
SBFspot_SRC_303_Linux_Win32

Dont know why the formating of the deletet line is so ugly. Sorry for that.

**Comment by SBFspot** [comment]
[SBF@27/12/2014]
You're right ;-)
When all values=0 (before actual start of inverter) it goes indeed wrong. I tried it with this little piece of code:
``` C
#include <iostream>

int main(int argc, char* argv[])
{

	int arr[] = {0,0,0,0,0,0,0,0,0,0};
	unsigned int idx;
	unsigned int first_rec, last_rec;
	
	for (first_rec = 0; (arr[first_rec] <= 0) && (first_rec < sizeof(arr)/sizeof(arr[0])); first_rec++);
	for (last_rec = (sizeof(arr)/sizeof(arr[0]))-1; (arr[last_rec] <= 0) && (last_rec >= 0); last_rec--);

	for (idx = first_rec-1; idx < last_rec+1; idx++)
		std::cout << arr[idx] << std::endl;

	return 0;
}
```

I'll try to fix this asap

**Comment by SBFspot** [comment]
[janus4@27/12/2014]
Hi SBF,
thank you for your efort.

So i can throw away the debug output i created 5 minutes ago.... :-)

Greetings

Mario

**Comment by SBFspot** [comment]
[SBF@28/12/2014]
Can you test with this file? Attachment at the top of this page.

**Comment by SBFspot** [comment]
[janus4@30/12/2014]
Will do a test tonight with the new File.

**Comment by SBFspot** [comment]
[janus4@30/12/2014]
Hi SBF,
as i had no data today (Snow on the panels) i have run SBFspot with the new .cpp file. No SIGSEGV bevore midnight and after midnight also no SIGSEGV.

Greetings

Mario

**Comment by SBFspot** [comment]
[janus4@30/12/2014]
One Question: Have you thought about altering the code in the SQLite code also? I dont use SQLite but i think the bug is(was?) there, too.

Greetings

Mario

**Comment by SBFspot** [comment]
[SBF@31/12/2014]
Yesterday I released V3.0.5 - issue solved for both MySQL and SQLite
Thanks for the feedback

**Comment by SBFspot** [comment]
**Issue closed by _SBF_ with comment**
Fixed in V3.0.5

**Reason closed**
Fixed
